File: app/src/main/java/com/botijasoftware/utils/GLMatrixStack.java
package com.botijasoftware.utils;

public class GLMatrixStack {
	public int MAX_STACK = 12;
	private GLMatrix[] stack = new GLMatrix[MAX_STACK];
	private int index;
	
	public GLMatrixStack() {
		for (int i=0; i<MAX_STACK;i++) {
			stack[i] = new GLMatrix();
		}
		index = 0;
	}
	
	public void pushMatrix(GLMatrix m) {
		if (index < MAX_STACK) {
			stack[index++].restore(m);
		}
	}
	
	public void popMatrix(GLMatrix m) {
		if (index > 0) {
			stack[--index].save(m);
		}
	}
}

File: app/src/main/java/com/botijasoftware/utils/GLMatrixStack.kt
package com.botijasoftware.utils

class GLMatrixStack {
    var MAX_STACK = 12
    private val stack = arrayOfNulls<GLMatrix>(MAX_STACK)
    private var index: Int = 0

    init {
        for (i in 0..MAX_STACK - 1) {
            stack[i] = GLMatrix()
        }
        index = 0
    }

    fun pushMatrix(m: GLMatrix) {
        if (index < MAX_STACK) {
            stack[index++]!!.restore(m)
        }
    }

    fun popMatrix(m: GLMatrix) {
        if (index > 0) {
            stack[--index]!!.save(m)
        }
    }
}